registry/remote/repository.go

```diff
@@ -895,7 +896,7 @@ func (s *manifestStore) Resolve(ctx context.Context, reference string) (ocispec.
 
 	switch resp.StatusCode {
 	case http.StatusOK:
-		return s.generateDescriptor(resp, ref)
+		return s.generateDescriptor(resp, ref, req.Method != "HEAD")
```

Comment:
just for my knowledge, if the request method is `HEAD`, will the server always return the digest in the header? what if the request is for a tag and the server doesn't return the `Docker-Content-Digest` header? we wouldn't be able to generate a descriptor at all in that case, would we?

Reply:
According to these test results, it will return that header in both cases; i.e., regardless of whether `image` is composed of a `tag`, or `digest`.  This is where I get thin on confidence though, so maybe you're seeing something I'm not.

The 4 cases below are labelled; 2 x pair of request/response, first via `digest` reference, second via `tag` reference.

```
% make main
[INFO] sourcing makefile.d/issue-225.mk
[INFO] sourced makefile.d/issue-225.mk
# >>> Request HTTP Headers [Digest Reference](https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:c84f0f11f3cdc4066362dc830e59f9d4b6a422eaaa94c2df1d3e7bfcb0384840)
#------------------------------------------------------------------------------- issue-225:main
% curl -sL --user AWS:${AWS_PWD} https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:c84f0f11f3cdc4066362dc830e59f9d4b6a422eaaa94c2df1d3e7bfcb0384840 -D- -o /dev/null
HTTP/1.1 200 OK
Content-Type: application/vnd.docker.distribution.manifest.v2+json
Docker-Distribution-Api-Version: registry/2.0
Sizes:
Date: Thu, 18 Aug 2022 19:25:43 GMT
Content-Length: 945


# <<< Response HTTP Headers [Digest Reference](https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:c84f0f11f3cdc4066362dc830e59f9d4b6a422eaaa94c2df1d3e7bfcb0384840)
#------------------------------------------------------------------------------- issue-225:main
% curl -sL --user AWS:${AWS_PWD} https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:c84f0f11f3cdc4066362dc830e59f9d4b6a422eaaa94c2df1d3e7bfcb0384840 -I
HTTP/1.1 200 OK
Content-Length: 945
Content-Type: application/vnd.docker.distribution.manifest.v2+json
Docker-Content-Digest: sha256:c84f0f11f3cdc4066362dc830e59f9d4b6a422eaaa94c2df1d3e7bfcb0384840
Docker-Distribution-Api-Version: registry/2.0
Sizes:
Date: Thu, 18 Aug 2022 19:25:43 GMT


# >>> Request HTTP Headers [Tag Reference](https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag)
#------------------------------------------------------------------------------- issue-225:main
% curl -sL --user AWS:${AWS_PWD} https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag -D- -o/dev/null
HTTP/1.1 200 OK
Content-Type: application/vnd.docker.distribution.manifest.v2+json
Docker-Distribution-Api-Version: registry/2.0
Sizes:
Date: Thu, 18 Aug 2022 19:25:43 GMT
Content-Length: 945


# <<< Response HTTP Headers [Tag Reference](https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag)
#------------------------------------------------------------------------------- issue-225:main
% curl -sL --user AWS:${AWS_PWD} https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag -I
HTTP/1.1 200 OK
Content-Length: 945
Content-Type: application/vnd.docker.distribution.manifest.v2+json
Docker-Content-Digest: sha256:c84f0f11f3cdc4066362dc830e59f9d4b6a422eaaa94c2df1d3e7bfcb0384840
Docker-Distribution-Api-Version: registry/2.0
Sizes:
Date: Thu, 18 Aug 2022 19:25:43 GMT
```

Reply:
I'm a bit confused here myself, I raised this question internally (20th of July) but haven't discovered anything new:
> Hmm, looks like the digest is returned from the server even if the reference is a tag (containing no digest)…

Now I'm trying to see exactly what HTTP call `oras pull 080565210187.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag` translates to which results in the header to be missing and the error `empty response Docker-Content-Digest`.

@Wwwsylvia do you know of an easier way to translate an oras call to its component HTTP request?   Meanwhile I can see if I can run the `cli` through a proxy of some sort and inspect the HTTP request headers that way.

Reply:
Okay here's a less ambiguous repro (part 1 of 2 from my last comment):
```
% make main
including makefile.d/headers.mk
included makefile.d/headers.mk
# HTTP Requests & Responses via [Tag Reference](https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag)
# >>> HEAD Request
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sILo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag |& grep -E "^[>]" | sed -E "s/(Authorization:.*) .*/\1 <REDACTED>/"
> HEAD /v2/my.repo/manifests/my.tag HTTP/1.1
> Host: 080565210187.dkr.ecr.us-west-2.amazonaws.com
> Authorization: Basic <REDACTED>
> User-Agent: curl/7.79.1
> Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json
>

# >>> HEAD Response
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sILo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag |& grep -E "^[<]"
< HTTP/1.1 200 OK
< Content-Length: 1151
< Content-Type: application/vnd.docker.distribution.manifest.v2+json
< Docker-Content-Digest: sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e
< Docker-Distribution-Api-Version: registry/2.0
< Sizes:
< Date: Sat, 20 Aug 2022 00:14:25 GMT
<

# >>> GET Request
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sLo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag |& grep -E "^[>]" | sed -E "s/(Authorization:.*) .*/\1 <REDACTED>/"
> GET /v2/my.repo/manifests/my.tag HTTP/1.1
> Host: 080565210187.dkr.ecr.us-west-2.amazonaws.com
> Authorization: Basic <REDACTED>
> User-Agent: curl/7.79.1
> Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json
>

# >>> GET Response
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sLo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag |& grep -E "^[<]"
< HTTP/1.1 200 OK
< Content-Type: application/vnd.docker.distribution.manifest.v2+json
< Docker-Distribution-Api-Version: registry/2.0
< Sizes:
< Date: Sat, 20 Aug 2022 00:14:26 GMT
< Content-Length: 1151
<

#
# HTTP Requests & Responses via [Digest Reference](https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e)
# >>> HEAD Request
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sILo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e |& grep -E "^[>]" | sed -E "s/(Authorization:.*) .*/\1 <REDACTED>/"
> HEAD /v2/my.repo/manifests/sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e HTTP/1.1
> Host: 080565210187.dkr.ecr.us-west-2.amazonaws.com
> Authorization: Basic <REDACTED>
> User-Agent: curl/7.79.1
> Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json
>

# >>> HEAD Response
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sILo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e |& grep -E "^[<]"
< HTTP/1.1 200 OK
< Content-Length: 1151
< Content-Type: application/vnd.docker.distribution.manifest.v2+json
< Docker-Content-Digest: sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e
< Docker-Distribution-Api-Version: registry/2.0
< Sizes:
< Date: Sat, 20 Aug 2022 00:14:26 GMT
<

# >>> GET Request
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sLo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e |& grep -E "^[>]" | sed -E "s/(Authorization:.*) .*/\1 <REDACTED>/"
> GET /v2/my.repo/manifests/sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e HTTP/1.1
> Host: 080565210187.dkr.ecr.us-west-2.amazonaws.com
> Authorization: Basic <REDACTED>
> User-Agent: curl/7.79.1
> Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json
>

# >>> GET Response
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sLo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e |& grep -E "^[<]"
< HTTP/1.1 200 OK
< Content-Type: application/vnd.docker.distribution.manifest.v2+json
< Docker-Distribution-Api-Version: registry/2.0
< Sizes:
< Date: Sat, 20 Aug 2022 00:14:27 GMT
< Content-Length: 1151
<

#
# Repro of issues/225 (HTTP GET)
#------------------------------------------------------------------------------- headers:main
% oras-main pull 080565210187.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag
########################## REQ #################################
GET /v2/my.repo/manifests/my.tag HTTP/1.1
Host: 080565210187.dkr.ecr.us-west-2.amazonaws.com
Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json


########################## RES #################################
HTTP/1.1 200 OK
Content-Length: 1151
Content-Type: application/vnd.docker.distribution.manifest.v2+json
Date: Sat, 20 Aug 2022 00:14:29 GMT
Docker-Distribution-Api-Version: registry/2.0
Sizes:


################################################################
Error: GET "https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag": empty response Docker-Content-Digest

make: *** [main] Error 1
%
```

Reply:
Got it!